File: nodejs/src/protocol.ts
import { numberToTuple, numberFromTuple } from './number.js';

export function createPacket(data: Uint8Array): Uint8Array {
  const retValue = new Uint8Array(data.byteLength + 4);
  const length = numberToTuple(data.byteLength);
  retValue[0] = length[0];
  retValue[1] = length[1];
  retValue[2] = length[2];
  retValue[3] = length[3];
  retValue.set(data, 4);
  return retValue;
}

export function extractData(packet: Uint8Array): Uint8Array {
  const header = new Uint8Array(packet.buffer, 0, 4);
  const len = numberFromTuple([header[0], header[1], header[2], header[3]]);
  if (len > packet.byteLength) throw new Error('packet length is greater then body');
  const body = new Uint8Array(packet.buffer, 4, len);
  return body;
}

File: rust/src/protocol.rs
use crate::number::{number_from_tuple, number_to_tuple};
use std::default::Default;
use std::io::{Read, Result, Stdin, Stdout, Write};
use std::vec::Vec;

pub fn read_packet(input: &mut Stdin) -> Result<Vec<u8>> {
    let mut header: [u8; 4] = [0; 4];
    input.read_exact(&mut header)?;
    let length = number_from_tuple(&header);
    let mut body = Vec::<u8>::new();
    body.resize_with(length as usize, Default::default);
    input.read_exact(&mut body)?;
    return Ok(body);
}

pub fn write_packet(output: &mut Stdout, data: &Vec<u8>) -> Result<()> {
    let len = data.len();
    let header = number_to_tuple(len as u32);
    output.write_all(&header)?;
    output.write_all(data)?;
    output.flush()?;
    Ok(())
}
